cmd/convertor/builder/builder_engine.go

```diff
@@ -0,0 +1,169 @@
+/*
+   Copyright The Accelerated Container Image Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package builder
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/containerd/containerd/archive/compression"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type BuilderEngineType int
+
+const (
+	BuilderEngineTypeOverlayBD BuilderEngineType = iota
+	BuilderEngineTypeFastOCI
+)
+
+type builderEngine interface {
+	DownloadLayer(ctx context.Context, idx int) error
+
+	// build layer archive, maybe tgz or zfile
+	BuildLayer(ctx context.Context, idx int) error
+
+	UploadLayer(ctx context.Context, idx int) error
+
+	// UploadImage upload new manifest and config
+	UploadImage(ctx context.Context) error
+
+	// cleanup remove workdir
+	Cleanup()
+}
+
+type builderEngineBase struct {
+	fetcher  remotes.Fetcher
+	pusher   remotes.Pusher
+	manifest specs.Manifest
+	config   specs.Image
+	workDir  string
+	oci      bool
+}
+
+func (e *builderEngineBase) isGzipLayer(ctx context.Context, idx int) (bool, error) {
+	rc, err := e.fetcher.Fetch(ctx, e.manifest.Layers[idx])
+	if err != nil {
+		return false, errors.Wrapf(err, "isGzipLayer: failed to open layer %d", idx)
+	}
+	drc, err := compression.DecompressStream(rc)
+	if err != nil {
+		return false, errors.Wrapf(err, "isGzipLayer: failed to open decompress stream for layer %d", idx)
+	}
+	compress := drc.GetCompression()
+	switch compress {
+	case compression.Uncompressed:
+		return false, nil
+	case compression.Gzip:
+		return true, nil
+	default:
+		return false, fmt.Errorf("unsupported layer format with compression %s", compress.Extension())
```

Comment:
other errors in this function start with "isGzipLayer:" this is not a consistent practice throughout the code, but I don't think it's bad to include the function name, should this last return also include it?